src/librustc_parse/parser/path.rs

```diff
@@ -71,7 +71,23 @@ impl<'a> Parser<'a> {
             debug!("parse_qpath: (decrement) count={:?}", self.unmatched_angle_bracket_count);
         }
 
-        self.expect(&token::ModSep)?;
+        let lo_colon = self.token.span;
+        if self.eat(&token::Colon) {
+            // <Bar as Baz<T>>:Qux
+            //                ^
+            let span = lo_colon.to(self.prev_span);
+            self.diagnostic()
+                .struct_span_err(span, "found single colon where type path was expected")
```

Comment:
```suggestion
                .struct_span_err(span, "found single colon before projection in qualified path")
```

Reply:
Because of this wording change, I would also add a check that `self.look_ahead(1, |t| t.is_ident())` to make sure that we're not dealing with code that is too far off from what we expect.

Reply:
I would use `self.look_ahead(1, |t| t.is_ident() && !t.is_reserved_ident())` just to be sure that it's not a keyword identifier after `:`.

Reply:
Just to clarify my understanding here - am I right in saying that this is a QOI (or quality of output) issue where we don't want to emit this specific error if the code we're parsing isn't followed by an identifier, because in that case we're only likely to confuse matters?

(I've assumed this with my latest change - noting it in the doc comments. Let me know if this is incorrect)

Reply:
Yeah that's the idea; basically we improve the heuristic to more narrowly target the specific mistake you wanted to provide suggestions for.